waywiser_twist_safety/src/rclcpp_components/emergency_stop_monitor_core.cpp

```diff
@@ -59,12 +65,39 @@ void EmergencyStopMonitor::emergency_stop_callback(
 
 void EmergencyStopMonitor::twist_callback(const geometry_msgs::msg::Twist::SharedPtr twist_msg)
 {
+  last_twist_msg_received_time_ = this->now();
   if (emergency_stop_value_) {
     twist_msg->linear.x = 0.0;
     twist_msg->angular.z = 0.0;
   }
   twist_publisher_->publish(*twist_msg);
 }
+
+void EmergencyStopMonitor::twist_watchdog_callback()
+{
+  if (last_twist_msg_received_time_.seconds() > 0) {
+    // Calculate the time elapsed since the last message was received
+    auto time_since_last_msg = (this->now() - last_twist_msg_received_time_).seconds();
+    if (time_since_last_msg > cmd_vel_in_timeout_) {
```

Comment:
As far as I can see, cmd_vel_in will not timeout after cmd_vel_in_timeout_ seconds but up to 2*cmd_vel_in_timeout_ seconds as cmd_vel_in_timeout_ is also the resolution of twist_watchdog_timer_ which checks for this condition.

I would suggest starting twist_watchdog_timer_ with the timeout cmd_vel_in_timeout_  as is but instead of calling a callback that periodically checks for (time_since_last_msg > cmd_vel_in_timeout_), the callback should only be called when there was no twist received in the meantime.
This can be achieved by changing l.68 to call setPeriod and reset the walltimer.

Reply:
I also updated similar functionality we have in waywiser_teleop/src/joy_emergency_stop.cpp accordingly.